### PR TITLE
Support Intel syntax in the assembly reports

### PR DESCRIPTION
**Pull request by wyk9787**
Fixes: #400

Add an option to support Intel syntax in the assembly report, which is used in both `disasm` and `weblist` formats.

Tested on both Mac OS and Linux.

```
(pprof) disasm absl::*
   ...
         .          .    28aa620: movl  %ebx, %r11d
   ...
(pprof) intel_syntax
(pprof) disasm absl::*
   ...
         .          .    28aa620: mov   r11d, ebx
   ...

```

### CODE REVIEW

**Comment by kalyanac** [comment]
Can you include test outputs?

File: internal/binutils/binutils_test.go [hunk 1]
```diff
@@ -188,10 +188,9 @@ func skipUnlessDarwinAmd64(t *testing.T) {
 	}
 }
 
-func TestDisasm(t *testing.T) {
-	skipUnlessLinuxAmd64(t)
+func DisasmTestingInternal(t *testing.T, intelSyntax bool) {
 	bu := &Binutils{}
-	insts, err := bu.Disasm(filepath.Join("testdata", "exe_linux_64"), 0, math.MaxUint64)
+	insts, err := bu.Disasm(filepath.Join("testdata", "exe_linux_64"), 0, math.MaxUint64, false)
```
Comment (kalyanac) [inline internal/binutils/binutils_test.go:193]: 'false' should be intelSyntax

File: internal/driver/commands.go [hunk 2]
```diff
@@ -154,6 +154,9 @@ var pprofVariables = variables{
 	"compact_labels": &variable{boolKind, "f", "", "Show minimal headers"},
 	"source_path":    &variable{stringKind, "", "", "Search path for source files"},
 	"trim_path":      &variable{stringKind, "", "", "Path to trim from source paths before search"},
+	"intel_syntax": &variable{boolKind, "f", "", helpText(
+		"Show assembly in Intel syntax",
+		"Only useful for command `disasm` and `weblist`")},
```
Comment (kalyanac) [inline internal/driver/commands.go:159]: Only applicable to `disasm` and `weblist`

File: internal/binutils/binutils_test.go (hunk 1 above)
Comment (wyk9787) [reply internal/binutils/binutils_test.go:193]: Do you mean I should switch the behavior for `Disasm`, i.e. passing `false` to `Disasm` to indicate users want Intel syntax?

File: driver/driver.go [hunk 3]
```diff
@@ -142,7 +142,7 @@ type ObjTool interface {
 
 	// Disasm disassembles the named object file, starting at
 	// the start address and stopping at (before) the end address.
-	Disasm(file string, start, end uint64) ([]Inst, error)
+	Disasm(file string, start, end uint64, intelSyntax bool) ([]Inst, error)
```
Comment (kalyanac) [inline driver/driver.go:145]: I wonder if accepting string is better to accommodate future changes in supported syntax. ex: asm_syntax=intel

File: internal/binutils/binutils_test.go [hunk 4]
```diff
@@ -188,10 +188,9 @@ func skipUnlessDarwinAmd64(t *testing.T) {
 	}
 }
 
-func TestDisasm(t *testing.T) {
-	skipUnlessLinuxAmd64(t)
+func DisasmTestingInternal(t *testing.T, intelSyntax bool) {
```
Comment (aalexand) [inline internal/binutils/binutils_test.go:191]: disasmTesting

Comment (aalexand) [reply internal/binutils/binutils_test.go:191]: Actually, testDisasm

File: internal/binutils/binutils.go [hunk 5]
```diff
@@ -157,12 +157,27 @@ func findExe(cmd string, paths []string) (string, bool) {
...
+				fmt.Sprintf("--start-address=%#x", start),
+				fmt.Sprintf("--stop-address=%#x", end),
+				file)
+		}
+	} else {
+		cmd = exec.Command(b.objdump, "-d", "-C", "--no-show-raw-insn", "-l",
+			fmt.Sprintf("--start-address=%#x", start),
+			fmt.Sprintf("--stop-address=%#x", end),
+			file)
+	}
```
Comment (aalexand) [inline internal/binutils/binutils.go:180]: Can we modify the arguments instead of duplicating the whole call? Make them an array, massage as needed, use ellipsis syntax to invoke exec.Command.

File: internal/driver/commands.go [hunk 6]
```diff
@@ -215,7 +218,7 @@ var pprofVariables = variables{
 		"For memory profiles, report average memory per allocation.",
 		"For time-based profiles, report average time per event.")},
 	"sample_index": &variable{stringKind, "", "", helpText(
-		"Sample value to report (0-based index or name)",
+		"Sample value to report (<x> is 0-based index or name)",
```
Comment (aalexand) [inline internal/driver/commands.go:221]: This doesn't seem related to this change.

File: internal/report/source.go [hunk 7]
```diff
@@ -239,7 +239,8 @@ func sourceCoordinates(asm map[int][]assemblyInstruction) (start, end int) {
 // assemblyPerSourceLine disassembles the binary containing a symbol
 // and classifies the assembly instructions according to its
 // corresponding source line, annotating them with a set of samples.
-func assemblyPerSourceLine(objSyms []*objSymbol, rs graph.Nodes, src string, obj plugin.ObjTool) map[int][]assemblyInstruction {
+func assemblyPerSourceLine(objSyms []*objSymbol, rs graph.Nodes,
+	src string, obj plugin.ObjTool, intelSyntax bool) map[int][]assemblyInstruction {
```
Comment (aalexand) [inline internal/report/source.go:243]: Please avoid unrelated formatting changes. Keep this one line, long lines are fine in Go.

**Comment by aalexand** [comment]
This PR seems to contain unrelated commits.

File: driver/driver.go (hunk 3 above)
Comment (aalexand) [reply driver/driver.go:145]: I wouldn't do it - this makes code harder to refactor. We can revise the current approach when we have more options. E.g. this could take an option structure of something. But just an argument is fine for now.

File: internal/driver/commands.go (hunk 2 above)
Comment (wyk9787) [reply internal/driver/commands.go]: Done.

File: driver/driver.go (hunk 3 above)
Comment (wyk9787) [reply driver/driver.go:145]: Ack.

File: internal/binutils/binutils_test.go (hunk 4 above)
Comment (wyk9787) [reply internal/binutils/binutils_test.go]: Done.

File: internal/binutils/binutils.go (hunk 5 above)
Comment (wyk9787) [reply internal/binutils/binutils.go]: Done.

File: internal/driver/commands.go (hunk 6 above)
Comment (wyk9787) [reply internal/driver/commands.go]: Reverted.

File: internal/report/source.go (hunk 7 above)
Comment (wyk9787) [reply internal/report/source.go]: Ack. Now fixed.

**Comment by wyk9787** [comment]
> This PR seems to contain unrelated commits.

Should now be reverted.

File: internal/report/report.go [hunk 8]
```diff
@@ -79,6 +79,8 @@ type Options struct {
 	Symbol     *regexp.Regexp // Symbols to include on disassembly report.
 	SourcePath string         // Search path for source files.
 	TrimPath   string         // Paths to trim from source file paths.
+
+	IntelSyntax bool // Whether or not to print assembly in Intel syntax
```
Comment (aalexand) [inline internal/report/report.go:83]: nit: missing trailing full stop.

Comment (wyk9787) [reply internal/report/report.go:83]: Done.

**Comment by wyk9787** [comment]
> to commands

Done.

File: internal/binutils/binutils_test.go (hunk 1 above)
Comment (wyk9787) [reply internal/binutils/binutils_test.go]: Now fixed.

**Comment by aalexand** [comment]
> Are you referring to "--start_address" and "--stop-address"? Do you mean they should be put after `file`?

I am referring to the file path argument (it's the only positional one: `--start-address` and `--stop-address` are not, these are options / flags).

**Comment by wyk9787** [comment]
> > Are you referring to "--start_address" and "--stop-address"? Do you mean they should be put after `file`?
> 
> I am referring to the file path argument (it's the only positional one: `--start-address` and `--stop-address` are not, these are options / flags).

I see. Right now, I am actually inserting the "intel flags" at the beginning of the `args`, so the file path argument is indeed the last argument.

**Comment by aalexand** [comment]
> I see. Right now, I am actually inserting the "intel flags" at the beginning of the `args`, so the file path argument is indeed the last argument.

Ah, I missed that, sorry.

File: internal/binutils/binutils.go [hunk 9]
```diff
@@ -157,12 +157,22 @@ func findExe(cmd string, paths []string) (string, bool) {
 
 // Disasm returns the assembly instructions for the specified address range
 // of a binary.
-func (bu *Binutils) Disasm(file string, start, end uint64) ([]plugin.Inst, error) {
+func (bu *Binutils) Disasm(file string, start, end uint64, intelSyntax bool) ([]plugin.Inst, error) {
 	b := bu.get()
-	cmd := exec.Command(b.objdump, "-d", "-C", "--no-show-raw-insn", "-l",
+	var cmd *exec.Cmd
```
Comment (aalexand) [inline internal/binutils/binutils.go:162]: Remove this var, and use "cmd :=" below, as prior to to this change.

File: internal/binutils/binutils.go [hunk 10]
```diff
@@ -157,12 +157,22 @@ func findExe(cmd string, paths []string) (string, bool) {
...
+	args := []string{"-d", "-C", "--no-show-raw-insn", "-l",
 		fmt.Sprintf("--start-address=%#x", start),
 		fmt.Sprintf("--stop-address=%#x", end),
-		file)
+		file}
+
+	if intelSyntax {
+		if runtime.GOOS == "darwin" {
+			args = append([]string{"-x86-asm-syntax=intel"}, args...)
```
Comment (aalexand) [inline internal/binutils/binutils.go:170]: nit: I personally would append this flag at the end, and move appending file to after this code. It doesn't really matter much in this case, but appending "args..." is O(N), unnecessarily. Again, this doesn't really matter here since the list is fixed len and tiny. So feel free to leave this as is.

File: internal/binutils/binutils.go (hunk 9 above)
Comment (wyk9787) [reply internal/binutils/binutils.go]: Done.

File: internal/binutils/binutils.go (hunk 10 above)
Comment (wyk9787) [reply internal/binutils/binutils.go]: It makes sense. Now changed.
